src/v/cluster/rm_stm.cc

```diff
@@ -2605,4 +2606,23 @@ ss::future<> rm_stm::handle_eviction() {
       });
 }
 
+ss::future<> rm_stm::clear_old_tx_pids() {
+    std::vector<model::producer_identity> pids_for_delete;
+    for (auto [id, epoch] : _log_state.fence_pid_epoch) {
+        auto pid = model::producer_identity(id, epoch);
+        // If pid is not inside tx_seqs it means we do not have transaction for
+        // it right now
+        if (!_log_state.tx_seqs.contains(pid)) {
+            pids_for_delete.push_back(pid);
+        }
+    }
+
+    co_await ss::max_concurrent_for_each(
+      pids_for_delete, 32, [this](auto pid) -> ss::future<> {
+          _mem_state.forget(pid);
```

Comment:
we need to do double check: one before outside `co_await` (you already have it) and another inside the lambda or use locks otherwise the state may change before it runs